### PR TITLE
REST v3 integration tests

### PR DESCRIPTION
**Pull request by jyotimahapatra**
Last PR for adding v3 api. Adds the REST v3 integration tests.

### CODE REVIEW

**Comment by jyotimahapatra** [comment]
@kyessenov Is it possible to get in a more recent version of envoy in gcr.io/istio-testing/go-control-plane-ci ?

The envoy binary is including `build_version` in the v3.Node 

```
{
  "node": {
    "id": "test-id",
    "cluster": "test-cluster",
    "build_version": "bb7ceff4c3c5bd4555dff28b6e56d27f2f8be0a7/1.13.0/Clean/RELEASE/BoringSSL",
    "user_agent_name": "envoy",
    "user_agent_build_version": {
      "version": {
        "major_number": 1,
        "minor_number": 13
      },
      "metadata": {
        "ssl.version": "BoringSSL",
        "revision.sha": "bb7ceff4c3c5bd4555dff28b6e56d27f2f8be0a7",
        "revision.status": "Clean",
        "build.type": "RELEASE"
      }
    },
    "extensions": [
      {
        "name": "envoy.filters.thrift.rate_limit",
        "category": "envoy.thrift_proxy.filters"
      },
      {
        "name": "envoy.filters.thrift.router",
        "category": "envoy.thrift_proxy.filters"
      },
      {
        "name": "envoy.retry_host_predicates.omit_canary_hosts",
        "category": "envoy.retry_host_predicates"
      },
      {
        "name": "envoy.retry_host_predicates.previous_hosts",
        "category": "envoy.retry_host_predicates"
      },
      {
        "name": "default",
        "category": "envoy.dubbo_proxy.route_matchers"
      },
      {
        "name": "envoy.transport_sockets.alts",
        "category": "envoy.transport_sockets.upstream"
      },
      {
        "name": "envoy.transport_sockets.raw_buffer",
        "category": "envoy.transport_sockets.upstream"
      },
      {
        "name": "envoy.transport_sockets.tap",
        "category": "envoy.transport_sockets.upstream"
      },
      {
        "name": "envoy.transport_sockets.tls",
        "category": "envoy.transport_sockets.upstream"
      },
      {
        "name": "raw_buffer",
        "category": "envoy.transport_sockets.upstream"
      },
      {
        "name": "tls",
        "category": "envoy.transport_sockets.upstream"
      },
      {
        "name": "envoy.health_checkers.redis",
        "category": "envoy.health_checkers"
      },
      {
        "name": "raw_udp_listener",
        "category": "envoy.udp_listeners"
      },
      {
        "name": "envoy.transport_sockets.alts",
        "category": "envoy.transport_sockets.downstream"
      },
      {
        "name": "envoy.transport_sockets.raw_buffer",
        "category": "envoy.transport_sockets.downstream"
      },
      {
        "name": "envoy.transport_sockets.tap",
        "category": "envoy.transport_sockets.downstream"
      },
      {
        "name": "envoy.transport_sockets.tls",
        "category": "envoy.transport_sockets.downstream"
      },
      {
        "name": "raw_buffer",
        "category": "envoy.transport_sockets.downstream"
      },
      {
        "name": "tls",
        "category": "envoy.transport_sockets.downstream"
      },
      {
        "name": "envoy.client_ssl_auth",
        "category": "envoy.filters.network"
      },
      {
        "name": "envoy.echo",
        "category": "envoy.filters.network"
      },
      {
        "name": "envoy.ext_authz",
        "category": "envoy.filters.network"
      },
      {
        "name": "envoy.filters.network.dubbo_proxy",
        "category": "envoy.filters.network"
      },
      {
        "name": "envoy.filters.network.kafka_broker",
        "category": "envoy.filters.network"
      },
      {
        "name": "envoy.filters.network.local_ratelimit",
        "category": "envoy.filters.network"
      },
      {
        "name": "envoy.filters.network.mysql_proxy",
        "category": "envoy.filters.network"
      },
      {
        "name": "envoy.filters.network.rbac",
        "category": "envoy.filters.network"
      },
      {
        "name": "envoy.filters.network.sni_cluster",
        "category": "envoy.filters.network"
      },
      {
        "name": "envoy.filters.network.thrift_proxy",
        "category": "envoy.filters.network"
      },
      {
        "name": "envoy.filters.network.zookeeper_proxy",
        "category": "envoy.filters.network"
      },
      {
        "name": "envoy.http_connection_manager",
        "category": "envoy.filters.network"
      },
      {
        "name": "envoy.mongo_proxy",
        "category": "envoy.filters.network"
      },
      {
        "name": "envoy.ratelimit",
        "category": "envoy.filters.network"
      },
      {
        "name": "envoy.redis_proxy",
        "category": "envoy.filters.network"
      },
      {
        "name": "envoy.tcp_proxy",
        "category": "envoy.filters.network"
      },
      {
        "name": "dubbo",
        "category": "envoy.dubbo_proxy.protocols"
      },
      {
        "name": "envoy.filters.dubbo.router",
        "category": "envoy.dubbo_proxy.filters"
      },
      {
        "name": "envoy.filters.udp_listener.udp_proxy",
        "category": "envoy.filters.udp_listener"
      },
      {
        "name": "auto",
        "category": "envoy.thrift_proxy.protocols"
      },
      {
        "name": "binary",
        "category": "envoy.thrift_proxy.protocols"
      },
      {
        "name": "binary/non-strict",
        "category": "envoy.thrift_proxy.protocols"
      },
      {
        "name": "compact",
        "category": "envoy.thrift_proxy.protocols"
      },
      {
        "name": "twitter",
        "category": "envoy.thrift_proxy.protocols"
      },
      {
        "name": "envoy.dog_statsd",
        "category": "envoy.stats_sinks"
      },
      {
        "name": "envoy.metrics_service",
        "category": "envoy.stats_sinks"
      },
      {
        "name": "envoy.stat_sinks.hystrix",
        "category": "envoy.stats_sinks"
      },
      {
        "name": "envoy.statsd",
        "category": "envoy.stats_sinks"
      },
      {
        "name": "envoy.dynamic.ot",
        "category": "envoy.tracers"
      },
      {
        "name": "envoy.lightstep",
        "category": "envoy.tracers"
      },
      {
        "name": "envoy.tracers.datadog",
        "category": "envoy.tracers"
      },
      {
        "name": "envoy.tracers.opencensus",
        "category": "envoy.tracers"
      },
      {
        "name": "envoy.tracers.xray",
        "category": "envoy.tracers"
      },
      {
        "name": "envoy.zipkin",
        "category": "envoy.tracers"
      },
      {
        "name": "envoy.retry_priorities.previous_priorities",
        "category": "envoy.retry_priorities"
      },
      {
        "name": "envoy.buffer",
        "category": "envoy.filters.http"
      },
      {
        "name": "envoy.cors",
        "category": "envoy.filters.http"
      },
      {
        "name": "envoy.csrf",
        "category": "envoy.filters.http"
      },
      {
        "name": "envoy.ext_authz",
        "category": "envoy.filters.http"
      },
      {
        "name": "envoy.fault",
        "category": "envoy.filters.http"
      },
      {
        "name": "envoy.filters.http.adaptive_concurrency",
        "category": "envoy.filters.http"
      },
      {
        "name": "envoy.filters.http.dynamic_forward_proxy",
        "category": "envoy.filters.http"
      },
      {
        "name": "envoy.filters.http.grpc_http1_reverse_bridge",
        "category": "envoy.filters.http"
      },
      {
        "name": "envoy.filters.http.grpc_stats",
        "category": "envoy.filters.http"
      },
      {
        "name": "envoy.filters.http.header_to_metadata",
        "category": "envoy.filters.http"
      },
      {
        "name": "envoy.filters.http.jwt_authn",
        "category": "envoy.filters.http"
      },
      {
        "name": "envoy.filters.http.on_demand",
        "category": "envoy.filters.http"
      },
      {
        "name": "envoy.filters.http.original_src",
        "category": "envoy.filters.http"
      },
      {
        "name": "envoy.filters.http.rbac",
        "category": "envoy.filters.http"
      },
      {
        "name": "envoy.filters.http.tap",
        "category": "envoy.filters.http"
      },
      {
        "name": "envoy.grpc_http1_bridge",
        "category": "envoy.filters.http"
      },
      {
        "name": "envoy.grpc_json_transcoder",
        "category": "envoy.filters.http"
      },
      {
        "name": "envoy.grpc_web",
        "category": "envoy.filters.http"
      },
      {
        "name": "envoy.gzip",
        "category": "envoy.filters.http"
      },
      {
        "name": "envoy.health_check",
        "category": "envoy.filters.http"
      },
      {
        "name": "envoy.http_dynamo_filter",
        "category": "envoy.filters.http"
      },
      {
        "name": "envoy.ip_tagging",
        "category": "envoy.filters.http"
      },
      {
        "name": "envoy.lua",
        "category": "envoy.filters.http"
      },
      {
        "name": "envoy.rate_limit",
        "category": "envoy.filters.http"
      },
      {
        "name": "envoy.router",
        "category": "envoy.filters.http"
      },
      {
        "name": "envoy.squash",
        "category": "envoy.filters.http"
      },
      {
        "name": "envoy.grpc_credentials.aws_iam",
        "category": "envoy.grpc_credentials"
      },
      {
        "name": "envoy.grpc_credentials.default",
        "category": "envoy.grpc_credentials"
      },
      {
        "name": "envoy.grpc_credentials.file_based_metadata",
        "category": "envoy.grpc_credentials"
      },
      {
        "name": "envoy.cluster.eds",
        "category": "envoy.clusters"
      },
      {
        "name": "envoy.cluster.logical_dns",
        "category": "envoy.clusters"
      },
      {
        "name": "envoy.cluster.original_dst",
        "category": "envoy.clusters"
      },
      {
        "name": "envoy.cluster.static",
        "category": "envoy.clusters"
      },
      {
        "name": "envoy.cluster.strict_dns",
        "category": "envoy.clusters"
      },
      {
        "name": "envoy.clusters.aggregate",
        "category": "envoy.clusters"
      },
      {
        "name": "envoy.clusters.dynamic_forward_proxy",
        "category": "envoy.clusters"
      },
      {
        "name": "envoy.clusters.redis",
        "category": "envoy.clusters"
      },
      {
        "name": "envoy.listener.http_inspector",
        "category": "envoy.filters.listener"
      },
      {
        "name": "envoy.listener.original_dst",
        "category": "envoy.filters.listener"
      },
      {
        "name": "envoy.listener.original_src",
        "category": "envoy.filters.listener"
      },
      {
        "name": "envoy.listener.proxy_protocol",
        "category": "envoy.filters.listener"
      },
      {
        "name": "envoy.listener.tls_inspector",
        "category": "envoy.filters.listener"
      },
      {
        "name": "envoy.file_access_log",
        "category": "envoy.access_loggers"
      },
      {
        "name": "envoy.http_grpc_access_log",
        "category": "envoy.access_loggers"
      },
      {
        "name": "envoy.tcp_grpc_access_log",
        "category": "envoy.access_loggers"
      },
      {
        "name": "dubbo.hessian2",
        "category": "envoy.dubbo_proxy.serializers"
      },
      {
        "name": "envoy.resource_monitors.fixed_heap",
        "category": "envoy.resource_monitors"
      },
      {
        "name": "envoy.resource_monitors.injected_resource",
        "category": "envoy.resource_monitors"
      },
      {
        "name": "envoy.ip",
        "category": "envoy.resolvers"
      },
      {
        "name": "auto",
        "category": "envoy.thrift_proxy.transports"
      },
      {
        "name": "framed",
        "category": "envoy.thrift_proxy.transports"
      },
      {
        "name": "header",
        "category": "envoy.thrift_proxy.transports"
      },
      {
        "name": "unframed",
        "category": "envoy.thrift_proxy.transports"
      }
    ]
  },
  "resource_names": [
    "cluster-v0-0"
  ],
  "type_url": "type.googleapis.com/envoy.config.endpoint.v3.ClusterLoadAssignment"
}
```

**Comment by kyessenov** [comment]
Yes, sure thing. Ideally, we would use some better docker registry than we have now, so that I'm not the sole point.

**Comment by jyotimahapatra** [comment]
@kyessenov Yes makes sense. @eapolinario Is working on the integration tests of [xds-relay](https://github.com/envoyproxy/xds-relay) and we are thinking about using a similar approach followed here. Is there a precedence of creating docker image in envoyproxy repo?

**Comment by kyessenov** [comment]
You should ask @mattklein123 how envoyproxy/envoy controls its build image access. I don't have visibility there.

**Comment by kyessenov** [comment]
https://github.com/envoyproxy/go-control-plane/pull/287
I don't know if 1.13.1 fixes the REST issue, but let's give it a try.

**Comment by jyotimahapatra** [comment]
Thanks @kyessenov for the prompt update.
Looks like `build_version` is still part of `DiscoveryRequest`. I will create an envoy bug for this.
```
{
  "node": {
    "id": "test-id",
    "cluster": "test-cluster",
    "build_version": "b67c14052c49890a7e3afe614d50979c346c024b/1.13.1/Clean/RELEASE/BoringSSL",
    "user_agent_name": "envoy",
    "user_agent_build_version": {
      "version": {
        "major_number": 1,
        "minor_number": 13,
        "patch": 1
      },
      "metadata": {
        "build.type": "RELEASE",
        "ssl.version": "BoringSSL",
        "revision.sha": "b67c14052c49890a7e3afe614d50979c346c024b",
        "revision.status": "Clean"
      }
    },
    "extensions": [
     ...
    ]
  },
  "resource_names": [
    "cluster-v0-2"
  ],
  "type_url": "type.googleapis.com/envoy.config.endpoint.v3.ClusterLoadAssignment"
}
```

File: pkg/server/v2/gateway.go [hunk 1]
```diff
@@ -57,27 +58,24 @@ func (h *HTTPGateway) ServeHTTP(resp http.ResponseWriter, req *http.Request) err
...
 	if err != nil {
-		http.Error(resp, "cannot read body", http.StatusBadRequest)
-		return nil
+		return nil, http.StatusBadRequest, fmt.Errorf("cannot read body")
 	}
 
 	// parse as JSON
 	out := &discovery.DiscoveryRequest{}
-	err = jsonpb.UnmarshalString(string(body), out)
+	unmarshaller := jsonpb.Unmarshaler{AllowUnknownFields: true}
```
Comment (jyotimahapatra) [inline pkg/server/v2/gateway.go:75]: @kyessenov Do you think its acceptable to go with `jsonpb.Unmarshaler{AllowUnknownFields: true}` to get around the issue of unknown fields? I've created an issue in envoy https://github.com/envoyproxy/envoy/issues/10633, and we can remove this special case when the fix lands.

**Comment by jyotimahapatra** [comment]
@kyessenov I think i corrected json issue. This is ready to be reviewed now.

**Comment by jyotimahapatra** [comment]
@kyessenov PTAL

**Pull request merged**